File: src/core/allocator.py
"""Allocate VLESS clients and routing rules from user permissions.

Pure processor: no I/O, no UUID generation, no config mutation. Callers supply
user permissions, the list of inbounds, and a pre-generated UUID map; this
module returns data structures the caller can inject wherever it wants.

Note: the protocol check uses the inline literal "vless" (not a named constant)
because the check appears exactly once. If Xray ever introduces another
protocol that should be treated like VLESS, update this single string.
"""


def allocate(user_permissions, inbounds, outbound_tags, uuids):
    """Allocate clients per inbound and routing rules per outbound.

    Args:
        user_permissions: {username: {"allowed_inbounds": [...],
                                       "allowed_outbounds": [...]}}
        inbounds: [{"tag": str, "protocol": str}, ...]
                   Caller may filter to VLESS before passing; non-VLESS
                   tags are skipped here with a warning.
        outbound_tags: [str, ...]  outbound tag strings from the config
        uuids: {"username@outboundtag": "uuid-v4-string", ...}
               Caller-owned UUID store. Read-only here.

    Returns:
        clients_by_inbound: {inbound_tag: [{"id": uuid, "email": email}, ...]}
        routing_rules: [{"user": ["regexp:.*@<tag>$"], "outboundTag": tag}, ...]
        warnings: list of human-readable strings about skipped entries
    """
    protocols = {entry["tag"]: entry.get("protocol") for entry in inbounds}
    clients_by_inbound = {}
    warnings = []
    seen_warnings = set()

    for username, perms in user_permissions.items():
        for outbound_tag in perms.get("allowed_outbounds", []):
            email = "%s@%s" % (username, outbound_tag)
            client_id = uuids.get(email)
            if not client_id:
                message = "missing uuid for %r" % email
                if message not in seen_warnings:
                    seen_warnings.add(message)
                    warnings.append(message)
                continue
            for inbound_tag in perms.get("allowed_inbounds", []):
                if protocols.get(inbound_tag) != "vless":
                    message = "skipping non-vless inbound '%s'" % inbound_tag
                    if message not in seen_warnings:
                        seen_warnings.add(message)
                        warnings.append(message)
                    continue
                clients_by_inbound.setdefault(inbound_tag, []).append(
                    {"id": client_id, "email": email}
                )

    routing_rules = [
        {"user": ["regexp:.*@%s$" % tag], "outboundTag": tag}
        for tag in outbound_tags
    ]

    return clients_by_inbound, routing_rules, warnings

File: src/core/allocator.ts
/**
 * Allocate VLESS clients and routing rules from user permissions.
 *
 * WHY this exists: turning permissions into `settings.clients` and
 * `routing.rules` is a pure transformation — no I/O, no UUID generation,
 * no config mutation — so it runs identically anywhere and is unit-tested
 * with plain dicts. Callers supply the permissions, the inbound list, the
 * outbound tags, and a pre-generated email→uuid map.
 *
 * WHY the protocol check is the inline literal "vless": the check appears
 * exactly once, so a named constant would add indirection without adding
 * meaning. If Xray ever grows another VLESS-like protocol, this one string
 * is the place to change.
 */

import { pyReprString } from "./python_json";

/** WHAT: the projected permission map the allocator consumes. */
export type UserPermissions = Record<
  string,
  { allowed_inbounds?: string[]; allowed_outbounds?: string[] }
>;

/** WHAT: the inbound fields the allocator reads. */
export type InboundEntry = { tag: string; protocol?: string };

/** WHAT: one generated routing rule (matcher list plus outbound tag). */
export type RoutingRule = { user: string[]; outboundTag: string };

/** WHAT: clients grouped by inbound tag, in generation order. */
export type ClientsByInbound = Record<string, { id: string; email: string }[]>;

/**
 * WHAT: allocate clients per inbound and routing rules per outbound.
 *
 * WHY the loop order is user → outbound → inbound: a missing uuid is a
 * property of the `username@outbound` email, so it is warned once per
 * email and skips every inbound under that exit, exactly like the copied
 * Python core. Warnings are de-duplicated by text.
 */
export function allocate(
  userPermissions: UserPermissions,
  inbounds: InboundEntry[],
  outboundTags: string[],
  uuids: Record<string, string | undefined>,
): [ClientsByInbound, RoutingRule[], string[]] {
  const protocols: Record<string, unknown> = {};
  for (const entry of inbounds) {
    protocols[entry["tag"]] = entry["protocol"];
  }
  const clientsByInbound: ClientsByInbound = {};
  const warnings: string[] = [];
  const seenWarnings = new Set<string>();

  for (const [username, permissions] of Object.entries(userPermissions)) {
    for (const outboundTag of permissions["allowed_outbounds"] ?? []) {
      const email = `${username}@${outboundTag}`;
      const clientId = uuids[email];
      if (!clientId) {
        const message = `missing uuid for ${pyReprString(email)}`;
        if (!seenWarnings.has(message)) {
          seenWarnings.add(message);
          warnings.push(message);
        }
        continue;
      }
      for (const inboundTag of permissions["allowed_inbounds"] ?? []) {
        if (protocols[inboundTag] !== "vless") {
          const message = `skipping non-vless inbound '${inboundTag}'`;
          if (!seenWarnings.has(message)) {
            seenWarnings.add(message);
            warnings.push(message);
          }
          continue;
        }
        (clientsByInbound[inboundTag] ??= []).push({
          id: clientId,
          email,
        });
      }
    }
  }

  const routingRules: RoutingRule[] = outboundTags.map((tag) => ({
    user: [`regexp:.*@${tag}$`],
    outboundTag: tag,
  }));

  return [clientsByInbound, routingRules, warnings];
}
